File: 킬링캠프/1주 차 - 완전탐색, 백트래킹, 이분탐색/challenge/N-Queens/Java/Solution.java
import java.util.*;

class Solution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> result = new ArrayList<>();
        backTracking(n, new boolean[n][n], result, new ArrayList<>(), 0);

        return result;
    }

    private void backTracking(int n, boolean[][] queens, List<List<String>> result, List<String> temp, int rowStart) {
        if (temp.size() == n) {
            result.add(new ArrayList<>(temp));
            return;
        }

        for (int row = rowStart; row < queens.length; row++) {
            for (int col = 0; col < queens[row].length; col++) {
                if (!checkOtherQueens(queens, row, col)) {
                    continue;
                }

                queens[row][col] = true;
                String location = createLocation(n, col);
                temp.add(location);
                backTracking(n, queens, result, temp, row + 1);
                queens[row][col] = false;
                temp.remove(temp.size() - 1);
            }
        }
    }

    private boolean checkOtherQueens(boolean[][] queens, int targetRow, int targetCol) {
        for (int row = 0; row < queens.length; row++) {
            if (queens[row][targetCol]) {
                return false;
            }
        }

        int currentRow = targetRow;
        int currentCol = targetCol;
        while (true) {
            currentRow -= 1;
            currentCol -= 1;

            if (!isSafe(queens.length, currentRow, currentCol)) {
                break;
            }

            if (queens[currentRow][currentCol]) {
                return false;
            }
        }

        currentRow = targetRow;
        currentCol = targetCol;
        while (true) {
            currentRow += 1;
            currentCol += 1;

            if (!isSafe(queens.length, currentRow, currentCol)) {
                break;
            }

            if (queens[currentRow][currentCol]) {
                return false;
            }
        }

        currentRow = targetRow;
        currentCol = targetCol;
        while (true) {
            currentRow -= 1;
            currentCol += 1;

            if (!isSafe(queens.length, currentRow, currentCol)) {
                break;
            }

            if (queens[currentRow][currentCol]) {
                return false;
            }
        }

        currentRow = targetRow;
        currentCol = targetCol;
        while (true) {
            currentRow += 1;
            currentCol -= 1;

            if (!isSafe(queens.length, currentRow, currentCol)) {
                break;
            }

            if (queens[currentRow][currentCol]) {
                return false;
            }
        }

        return true;
    }

    private boolean isSafe(int n, int row, int col) {
        return row >= 0 && row < n && col >= 0 && col < n;
    }

    private String createLocation(int n, int col) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            if (i == col) {
                sb.append("Q");
            } else {
                sb.append(".");
            }
        }

        return sb.toString();
    }
}

File: 킬링캠프/1주 차 - 완전탐색, 백트래킹, 이분탐색/challenge/N-Queens/solution.py
class Solution:
    DIR = [[-1, 0], [0, 1], [0, -1], [1, 0], [-1, 1], [-1, -1], [1, 1], [1, -1]]  # 북, 동, 서, 남, 북동, 북서, 남동, 남서

    def solveNQueens(self, n: int) -> List[List[str]]:
        result = []
        visited = [[False for _ in range(n)] for _ in range(n)]
        self.back_tracking(n, 0, 0, result, visited, 1)

        return result
    
    def back_tracking(self, n, start_row, start_col, result, visited, step):
        if (step > n):
            result.append(self.convert_visited_to_result(visited))
            return
        
        for row in range(start_row, n):
            start_col = start_col if row == start_row else 0
            for col in range(start_col, n):
                if (not self.is_safe(row, col, n, visited)):
                    continue
                
                visited[row][col] = True
                next_col = (col + 1) % n
                next_row = row + 1 if (next_col == 0) else row
                self.back_tracking(n, next_row, next_col, result, visited, step + 1)

                visited[row][col] = False
    
    def is_safe(self, row, col, n, visited):
        for d in self.DIR:
            next_row = row + d[0]
            next_col = col + d[1]

            while (next_row >= 0 and next_row < n and next_col >= 0 and next_col < n):
                if (visited[next_row][next_col]):
                    return False
                
                next_row += d[0]
                next_col += d[1]
        
        return True
    
    def convert_visited_to_result(self, visited):
        result = []
        for row in visited:
            row_str = ""
            for col in row:
                row_str += "Q" if (col) else "."
            result.append(row_str)
        
        return result
